### PR TITLE
API: add CloseEvent

### PR DESCRIPTION
**Pull request by caugner**
[CloseEvent](https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent#Browser_compatibility)

### CODE REVIEW

File: api/CloseEvent.json [hunk 1]
```diff
@@ -0,0 +1,101 @@
+{
+  "api": {
+    "CloseEvent": {
+      "__compat": {
+        "mdn_url": "https://developer.mozilla.org/docs/Web/API/CloseEvent",
+        "support": {
+          "webview_android": {
+            "version_added": null
```
Comment (jpmedley) [inline api/CloseEvent.json]: All 3 versions of Chrome are `true`. You can find the complete list here: https://cs.chromium.org/chromium/src/third_party/blink/renderer/modules/websockets/close_event.idl

File: api/CloseEvent.json [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            "version_added": "8"
+          },
+          "ie": {
+            "version_added": "10"
+          },
+          "opera": {
+            "version_added": null
+          },
+          "opera_android": {
+            "version_added": null
```
Comment (jpmedley) [inline api/CloseEvent.json]: `true` for both versions of Opera.

File: api/CloseEvent.json [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+          "mdn_url": "https://developer.mozilla.org/docs/Web/API/CloseEvent/initCloseEvent",
+          "support": {
+            "webview_android": {
+              "version_added": null
+            },
+            "chrome": {
+              "version_added": false
+            },
+            "chrome_android": {
+              "version_added": null
```
Comment (jpmedley) [inline api/CloseEvent.json]: Ditto all 3. See above.

File: api/CloseEvent.json [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+              "version_added": "41"
+            },
+            "ie": {
+              "version_added": false
+            },
+            "opera": {
+              "version_added": false
+            },
+            "opera_android": {
+              "version_added": false
```
Comment (jpmedley) [inline api/CloseEvent.json]: Ditto both. See above.

File: api/CloseEvent.json [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+              "version_added": true
+            },
+            "edge_mobile": {
+              "version_added": true
+            },
+            "firefox": {
+              "version_added": "41"
+            },
+            "firefox_android": {
+              "version_added": "41"
```
[the remark this answers is not in the document]
Comment (caugner) [reply api/CloseEvent.json]: Ah, it was removed in FF 41, not added:
https://bugzilla.mozilla.org/show_bug.cgi?id=1161950

File: api/CloseEvent.json (hunk 3 above)
Comment (caugner) [reply api/CloseEvent.json]: @jpmedley According to the [Chromium Code Search](https://cs.chromium.org/search/?q=initCloseEvent&ssfr=1&type=cs), `initCloseEvent` isn't actually supported by Chrome. Can you elaborate on that?

File: api/CloseEvent.json (hunk 1 above)
Comment (caugner) [reply api/CloseEvent.json]: Fixed in c2d8949f0be1c60887c66b19230011ffdac21ffa.

File: api/CloseEvent.json (hunk 2 above)
Comment (caugner) [reply api/CloseEvent.json]: Fixed in c2d8949f0be1c60887c66b19230011ffdac21ffa.

File: api/CloseEvent.json (hunk 3 above)
Comment (caugner) [reply api/CloseEvent.json]: For the time being, I have set all to `false` in c2d8949f0be1c60887c66b19230011ffdac21ffa.

File: api/CloseEvent.json (hunk 4 above)
Comment (caugner) [reply api/CloseEvent.json]: For the time being, I have not changed these, see [above](https://github.com/mdn/browser-compat-data/pull/1767#discussion_r181196508).

**Comment by Elchi3** [comment]
This misses the constructor as a sub feature:
https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent/CloseEvent

**Comment by caugner** [comment]
@Elchi3 I have now added the constructor as a subfeature in 3c2e830aa757937d5f1d764694c2931caab5280c.

File: api/CloseEvent.json [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+          },
+          "edge": {
+            "version_added": true
+          },
+          "edge_mobile": {
+            "version_added": true
+          },
+          "firefox": {
+            "version_added": "8",
+            "notes": "Prior to Firefox 8, the WebSocket close event was sent to the listener as a simple event. Prior to Firefox 12, the close code CLOSE_NORMAL was used when the channel was closed due to an unexpected error or unspecified error conditoin."
```
Comment (ddbeck) [inline api/CloseEvent.json]: A few possible improvements to the notes here:

* Split "Prior to Firefox 8" and "Prior to Firefox 12" into separate notes (Firefox 12 note first)
* Code tags: `<code>CLOSE_NORMAL</code>`
* Typo: "conditoin" ➡️ "condition"

File: api/CloseEvent.json [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+            "version_added": true
+          },
+          "edge": {
+            "version_added": true
+          },
+          "edge_mobile": {
+            "version_added": true
+          },
+          "firefox": {
+            "version_added": "8",
```
Comment (ddbeck) [inline api/CloseEvent.json]: The note implies this should be some version before 8 or, if unknown, `true`

File: api/CloseEvent.json [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+            },
+            "opera_android": {
+              "version_added": null
+            },
+            "safari": {
+              "version_added": null
+            },
+            "safari_ios": {
+              "version_added": null
+            }
```
Comment (ddbeck) [inline api/CloseEvent.json]: Is there a reason there's no status information here?

File: api/CloseEvent.json (hunk 7 above)
Comment (caugner) [reply api/CloseEvent.json]: I think it's correct.

Prior to Firefox 8, Firefox did not support the `CloseEvent` as defined by the spec. Rather, all listeners registered via `WebSocket.onclose()` were notified with an unspecific event, presumably without the properties of a real `CloseEvent`.

One could argue though that this note would be better placed in `WebSocket/onclose`.

File: api/CloseEvent.json (hunk 8 above)
Comment (caugner) [reply api/CloseEvent.json]: Hm, I don't see any compat data for Safari on MDN, except for the `false` regarding `initCloseEvent`. Are you aware of other sources we could get this information from?

File: api/CloseEvent.json (hunk 6 above)
Comment (caugner) [reply api/CloseEvent.json]: Thanks, fixed in 20a7bf6 now.

File: api/CloseEvent.json (hunk 7 above)
Comment (ddbeck) [reply api/CloseEvent.json]: Ah, thank you for the explanation. I should've read it more closely—it makes sense your way.

File: api/CloseEvent.json (hunk 8 above)
Comment (ddbeck) [reply api/CloseEvent.json]: Sorry for the confusion. This wasn't a gripe about the Safari data. I meant this is a question about the `"status"` block (i.e., whether this is experimental, standard, etc.)

Comment (caugner) [reply api/CloseEvent.json]: Note to myself: The consistency checker (#2463) could check that if a parent feature has a status block, the child feature must also have a status block (and must be consistent: if the parent feature is experimental/deprecated, the child feature must also be experimental/deprecated; if the parent feature is not on a standard-track, the child feature must not be on a standard-track either).

Comment (ddbeck) [reply api/CloseEvent.json]: @caugner while this is in progress, is it OK if I continue to review other PRs from you? I don't want to bombard you with notifications, if that's going to be too noisy for you

Comment (caugner) [reply api/CloseEvent.json]: @ddbeck Sure, that's fine. (My only problem really is "reading" notifications on the go and not being able to mark them as "unread" again, which is the reason I didn't react here.)

Comment (ddbeck) [reply api/CloseEvent.json]: @caugner Yeah, I have the same problem! I get email notifications too, but that's it's own set of headaches

Comment (caugner) [reply api/CloseEvent.json]: @ddbeck have added the status now. :wink:

**Comment by ddbeck** [comment]
Thanks yet again, @caugner!

**Pull request merged**